### PR TITLE
Irsa

### PR DESCRIPTION
**Pull request by ddl-wadkars**
IRSA blog and hands on tutorial and runbook

### CODE REVIEW

File: .idea/.gitignore
Comment (vaibhav-dhawan) [inline .idea/.gitignore:1]: Can you remove the `.idea` folder from the commit (or add it to gitignore)

File: IRSA-README.md [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+and the AWS Administration teams?
+
+It is also possible that two teams working on separate Domino data planes hosted on different Cloud Providers in the
+same organization may need to consume resources from each other. Data planes hosted in each of these Cloud Providers 
+may need to assume identities in the other Cloud Providers.  
+
+Using Web Identities based on an OpenID Connect Provider supports such use-cases. In this article we will start with a 
+basic setup where using IAM Role for Service Accounts (IRSA) we can configure Kubernetes Service Accounts attached to 
+an EKS workload to assume IAM Role(s). This is similar to attaching an IAM Role to an EC2 instance at startup. 
+Note that this mechanism allows EKS workloads to assume roles in any AWS Account and not just the AWS account hosting
```
Comment (vaibhav-dhawan) [inline IRSA-README.md:31]: Lets throw in the 'cross-account IAM roles' terminology here?

File: IRSA-README.md [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+    trust the Domino Platform
+    
+  - IRSA support in Domino allows configuration of multiple roles for a Domino workload where the boto3 library can 
+    switch across the multiple profiles configured in the ~/.aws/config file
+    
+  - The mapping of a workload to a set of roles can be configured based on custom configurable rules. This allows for 
+    complex mappings to be supported
+
+### How does IRSA work with Domino?
+ 1 . For a Domino Workload (implemented as a K8s Pod) to be able to assume an IAM Role, it must use the
```
Comment (vaibhav-dhawan) [inline IRSA-README.md:63]: Formatting of the bullets is a bit inconsistent (spaces at the start) but thats a minor thing

File: IRSA-README.md [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+
+### How does IRSA work with Domino?
+ 1 . For a Domino Workload (implemented as a K8s Pod) to be able to assume an IAM Role, it must use the 
+ `STS:AssumeRoleWithWebIdentity` API which returns temporary security credentials. But first the caller (in this case 
+ a pod) needs to be authenticated with a Web Identity Provider that is compatible with the OpenID connect (OIDC) protocol.
+
+ 2. Starting [1.12](https://kubernetes.io/docs/concepts/storage/projected-volumes/) Kubernetes allows an OIDC provider
+    to be associated with it. An EKS Cluster provides a unique OIDC provider bound to the `KubeAPIServer` server. A JWT 
+    created by this OIDC provider can be mounted as a projected volume in the Pod. This JWT is associated with the K8s 
+    Service Account associated with the workload. It is distinct from the JWT issued by the `KubeAPIServer`. This JWT
```
Comment (vaibhav-dhawan) [inline IRSA-README.md:70]: `It is distinct from the JWT issued by the `KubeAPIServer`.` I'd leave this bit out, it can confuse people since technically both tokens are issued by the kube server, just the targets are different

File: IRSA-README.md [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+ a pod) needs to be authenticated with a Web Identity Provider that is compatible with the OpenID connect (OIDC) protocol.
+
+ 2. Starting [1.12](https://kubernetes.io/docs/concepts/storage/projected-volumes/) Kubernetes allows an OIDC provider
+    to be associated with it. An EKS Cluster provides a unique OIDC provider bound to the `KubeAPIServer` server. A JWT 
+    created by this OIDC provider can be mounted as a projected volume in the Pod. This JWT is associated with the K8s 
+    Service Account associated with the workload. It is distinct from the JWT issued by the `KubeAPIServer`. This JWT 
+    acts as a workload identity credential that can be validated by AWS STS in order to pass IAM role credentials to 
+    that workload.
+
+3. This provider is configured as an IAM identity Provider within an AWS Account the EKS workloads want to assume an
```
Comment (vaibhav-dhawan) [inline IRSA-README.md:74]: ` IAM identity Provider within an AWS Account the EKS workloads want to assume an `
Should reword this a bit, not clear

File: IRSA-README.md [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+
+3. This provider is configured as an IAM identity Provider within an AWS Account the EKS workloads want to assume an 
+   IAM Role. The key point here is that the EKS cluster and the IAM Role being assumed by a workload running on EKS do 
+   not need to be in the same AWS Account. 
+
+4. An AWS Admin will configure the IAM Role Trust Policy to enable a specific K8s Service Account (as identified by the
+   JWT above) to assume the Role. AWS STS will validate the Workload Identity and return temporary IAM Role Credentials
+   to a workload if an IAM Role Trust Policy allows it.
+
+5. Typically the amazon-eks-pod-identity-webhook will apply mutations based on annotations defined on the Service
```
Comment (vaibhav-dhawan) [inline IRSA-README.md:82]: Should we mention the aws webhook? Maybe we just say domino enables you to perform the same pattern on any non-aws kubernetes cluster 

Not a big deal, we can leave it in as well

File: advanced-credential-propagation/irsa/tutorial.md [hunk 6]
```diff
@@ -0,0 +1,432 @@
...
+kubectl -f domino-platform apply -f user-identity-base-irsa.yaml
+```
+
+This yaml defines a mutation using the `domsed` which is a Domino aware mutating webhook. It does the following:
+
+a. Update the user to K8s service account mappings (see Pre-requisites)
+
+Next update the user to K8s service account mappings (see Pre-requisites)
+
+```yaml
```
Comment (vaibhav-dhawan) [inline advanced-credential-propagation/irsa/tutorial.md:28]: Doesnt this conflict with the direction that the SA name should match the username? And we added a thing to the mutation to use this logic too right

File: IRSA-README.md (hunk 1 above)
Comment (ddl-wadkars) [reply IRSA-README.md:31]: Good idea I will rephrase it as - "This mechanism allows EKS workloads to assume IAM roles across AWS Accounts. You are not limited to assuming an IAM role in the AWS Account hosting the EKS cluster. Eventually through the series of articles we will discover how any K8s cluster hosted on premise or any cloud provider can assuming an IAM role in an AWS Account using this mechanism"

File: IRSA-README.md (hunk 2 above)
Comment (ddl-wadkars) [reply IRSA-README.md:63]: Fixed

File: IRSA-README.md (hunk 3 above)
Comment (ddl-wadkars) [reply IRSA-README.md:70]: Good point

File: IRSA-README.md (hunk 4 above)
Comment (ddl-wadkars) [reply IRSA-README.md:74]: How does this sound
"3. Next, the OIDC provider attached to the EKS cluster is configured as an Identity Provider within which the AWS Account  the Domino workloads want to assume an IAM Role. The key point is Domino workloads can assume IAM roles across multiple AWS Accounts and are not limited to doing to in the account hosting the EKS cluster."

File: IRSA-README.md (hunk 5 above)
Comment (ddl-wadkars) [reply IRSA-README.md:82]: Good point - I rewrote


5. Domino adaptation of IRSA operates on a shared trust policy. A Domino mutating webhook framework called [Domsed](https://github.com/dominodatalab/domino-field-solutions-installations/tree/main/domsed) 
   configures a Domino workload definition to enable passing an identity issued by the OIDC Provider to IAM. We will refer to this identity as the Workload Identity. An AWS Administrator on the other hand configures a trust relationship between an IAM Role and this Workload Identity. This allows a Domino workload to assume one or more roles if those IAM Roles have their Trust Policy configured to permit the workload identity to assume them.
   
 We will discover in subsequent articles of this series, how this mechanism is cloud provider neutral and allows workloads in Domino  data planes running on-premise or in other cloud providers like GCP and Azure to assume AWS IAM Roles.

File: advanced-credential-propagation/irsa/tutorial.md (hunk 6 above)
Comment (ddl-wadkars) [reply advanced-credential-propagation/irsa/tutorial.md:28]: I want to keep this general. We have the ability to auto-match but then I have to get into details like the naming conventions should not include a `@`. This is an explicit version . I can make a comment about the alternatives but in tutorials I think it would impact the flow of thought. What do you think

File: IRSA-README.md (hunk 4 above)
Comment (vaibhav-dhawan) [reply IRSA-README.md]: Sounds good!

File: advanced-credential-propagation/irsa/tutorial.md (hunk 6 above)
Comment (vaibhav-dhawan) [reply advanced-credential-propagation/irsa/tutorial.md:28]: Yea agreed lets keep it simple for the article i think, we dont have to go into every possible scenario

File: IRSA-README.md [hunk 7]
```diff
@@ -0,0 +1,88 @@
+# Domino Advanced Credential Propagation for AWS based on IAM Role for Service Accounts (IRSA)
+
+IAM Role for Service Accounts ([IRSA](https://docs.aws.amazon.com/eks/latest/userguide/iam-roles-for-service-accounts.html)) is a capability which allows Kubernetes Service Accounts attached to a EKS workload to assume IAM Role(s). In much the same way as you can attach and IAM Role to an EC2 instance at startup.
+
+## Article Topics
+ - Security 
+ - Identity And Access Management (IAM) 
+ - Role Based Access Control (RBAC)
+ - OpenID Connect (OIDC), Web Identity and Attribute Based Access Control (ABAC)
```
[the remark this answers is not in the document]
Comment (vaibhav-dhawan) [reply IRSA-README.md:9]: Not quite, the kubernetes SA token that is sent to AWS is an Auth token, not an access token, which is what makes it OAuth instead of OIDC. The terminology is confusing i agree :-) 
https://supertokens.com/blog/oauth-vs-oidc
`While OAuth describes what resources a user has access to, OIDC describes who the user is`
Basically any token that allows you to access a resource is an OAuth token. structurally they're both just JWTs though

[the remark this answers is not in the document]
Comment (ddl-wadkars) [reply IRSA-README.md:9]: Got it. So should I reword this as-
`OAuth, Web Identity and Attribute Based Access Control (ABAC) `
